lib/iris/experimental/ugrid/utils.py

```diff
@@ -0,0 +1,322 @@
+        # Transpose all inputs to mesh-last order.
+        tranpose_dims = [i_dim for i_dim in dim_range if i_dim != mesh_dim] + [
+            mesh_dim
+        ]  # chop out mesh_dim + put it at the end.
+
+        def transposed_copy(cube, dim_order):
+            cube = cube.copy()
+            cube.transpose()
+            return cube
+
+        full_mesh_cube = transposed_copy(full_mesh_cube, tranpose_dims)
+        region_cubes = [
+            transposed_copy(region_cube, tranpose_dims)
+            for region_cube in region_cubes
+        ]
+
+        # Also prepare for transforming the output back to the original order.
+        untranspose_dims = dim_range.copy()
+        # Neat trick to produce the reverse operation.
+        untranspose_dims[tranpose_dims] = dim_range
+
+    #
+    # Here's the core operation..
+    #
+    def fill_region(target, regiondata, regioninds):
+        if not target.flags.writeable:
+            # The initial input can be a section of a da.zeros(), which has no
+            # real array "behind" it.  This means that real arrays created in
+            # memory are only chunk-sized, but it also means that 'target' may
+            # not be writeable.  So take a copy to fix that, where needed.
+            target = target.copy()
+        # N.B. Indices are basically 1D, but may have leading *1 dims for
+        # alignment, to satisfy da.map_blocks
+        assert all(size == 1 for size in regioninds.shape[:-1])
+        inds = regioninds.flatten()
+        # Assign blocks with indexing on the last dim only.
+        target[..., inds] = regiondata
+        return target
+
+    # Create an initially 'empty' (all-masked) dask array matching the input.
+    # N.B. this does not use the full_mesh_cube.lazy_data() array, but only its
+    # shape and dtype, since the data itself is not used in the calculation.
+    # N.B. chunking matches the input cube, allowing performance control.
+    input_data = full_mesh_cube.lazy_data()
+    result_array = da.ma.masked_array(
+        da.zeros(
+            input_data.shape,
+            dtype=input_data.dtype,
+            chunks=input_data.chunksize,
+        ),
+        True,
+    )
+
+    # Wrap this repeatedly with a lazy operation to assign each region.
+    # It is done this way because we couldn't get map_blocks to correctly wrap
+    # a function which does all regions in a single operation.
+    # TODO: replace with a single-stage solution: Probably better, if possible.
+    # Notes on resultant calculation properties:
+    # 1. map_blocks is chunk-mapped, so it is parallelisable and space-saving
+    # 2. However, fetching less than a whole chunk is not efficient
+    for region_cube in region_cubes:
+        # Lazy data array from the region cube
+        datarr = region_cube.lazy_data()
+
+        # Lazy indices from the mesh-dim coord.
+        mesh_dimcoord = region_cube.coord(
+            name_or_coord=index_coord_name, dimensions=region_cube.ndim - 1
```

Comment:
@pp-mo Minor, but you could have something like an `rc_mesh_dim = region_cube.ndim - 1`, as you use it a couple of times in this loop i.e., `line+299` also

Reply:
Ok! :+1:

Reply:
Actually that's a bit horrible, because in this usage, we are using "cube.ndim - 1" as the "index of the last data dimension".
But in the next case below, we are using "cube.ndim - 1" as the "number of data dimensions, minus one", which we multiply by to get a (1,1,1,n) tuple.
So if you name this as 'mesh_dimension' (effectively, an index), then _multipying_ by it does look very odd !